cin_validator/rules/cin2022_23/rule_4016.py

```diff
@@ -0,0 +1,306 @@
+
+CINplanDates = CINTable.CINplanDates
+CINPlanStartDate = CINplanDates.CINPlanStartDate
+
+Header = CINTable.Header
+ReferenceDate = Header.ReferenceDate
+
+
+# define characteristics of rule
+@rule_definition(
+    # write the rule code here
+    code=4016,
+    # replace ChildProtectionPlans with the value in the module column of the excel sheet corresponding to this rule .
+    # Note that even if multiple tables are involved, one table will be named in the module column.
+    module=CINTable.CINplanDates,
+    # replace the message with the corresponding value for this rule, gotten from the excel sheet.
+    message="A CIN Plan has been reported as open at the same time as a Child Protection Plan.",
+    # The column names tend to be the words within the < > signs in the github issue description.
+    affected_fields=[
+        CINPlanStartDate,
+        CPPstartDate,
+        CPPendDate,
+    ],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    # PREPARING DATA
+
+    # Replace ChildProtectionPlans with the name of the table you need.
+    df_cpp = data_container[ChildProtectionPlans].copy()
+    df_cin = data_container[CINplanDates].copy()
+
+    # Before you begin, rename the index so that the initial row positions can be kept intact.
+    df_cpp.index.name = "ROW_ID"
+    df_cin.index.name = "ROW_ID"
+
+    # Resetting the index causes the ROW_IDs to become columns of their respective DataFrames
+    # so that they can come along when the merge is done.
+    df_cpp.reset_index(inplace=True)
+    df_cin.reset_index(inplace=True)
+
+    # ReferenceDate exists in the header table so we get header table too.
+    df_ref = data_container[Header]
+    ref_date_series = df_ref[ReferenceDate]
+
+    # the make_census_period function generates the start and end date so that you don't have to do it each time.
+    collection_start, reference_date = make_census_period(ref_date_series)
+
+    # lOGIC
+    # Implement rule logic as described by the Github issue.
+    # Put the description as a comment above the implementation as shown.
+
+    # The <CINPlanStartDate> (N00689) for any CIN Plan group cannot fall within either:
+    # <CPPstartDate> (N00105) or <CPPendDate> (N00115);
+    # or <CPPstartDate> and <ReferenceDate> (N00603) if <CPPendDate> is not present - for any CPP group;
+    # unless <CINPlanStartDate> is equal to <CPPendDate> for this group
+
+    #  Merge tables
+    df_merged = df_cin.merge(
+        df_cpp,
+        on=["LAchildID"],
+        how="left",
+        suffixes=("_cin", "_cpp"),
+    )
+
+    #  Get rows where CINPlanStartDate is between CPPstartDate and CPPendDate (or reference_date)
+    condition = (df_merged[CINPlanStartDate] >= df_merged[CPPstartDate]) & (
+        (df_merged[CINPlanStartDate] < df_merged[CPPendDate])
+        & df_merged[CPPendDate].notna()
+        | (df_merged[CINPlanStartDate] < reference_date) & df_merged[CPPendDate].isna()
```

Comment:
You remember in maths at school they always used to say, do one bit of the sum at a time so you can go back and see where you went wrong later, this feels a bit like that. It's taking me a little while to parse mentally.

Reply:
@SLornieCYC, I'm not entirely sure this works quite right, I think we need to separate some logical conditions with brackets a bit more to make sure it's doing what we think, or, we need to write them as separate conditions and combine them in the line that does the slicing. Even if it does work, the readability means that people who take over or help maintain the code in the future will struggle so we should refactor.

At the moment it checks if (CPPstart is between CINstart and CINend AND CINend is not NaN) or (CINstart is less than reference date) and (CPPend is NaN)

What we want to check is:
The <CINPlanStartDate> (N00689) for any CIN Plan group cannot fall within either:
<CPPstartDate> (N00105) or <CPPendDate> (N00115);
or <CPPstartDate> and <ReferenceDate> (N00603) if <CPPendDate> is not present - for any CPP group;
unless <CINPlanStartDate> is equal to <CPPendDate> for this group

Which is:
Fail if CPPstart < CINstart < CPPend
or fail if CPPstart < CINstart < RefDate if CPPend is not present
But allow both of these to pass is CINstart = CPPend for a group

What I think wouyld be better, especially as far as readability is something like 
condition 1 = CPPstart is between CINstart and CINend 
condition 2 =  CPPend isn't present and CPPstart is between CINstart and Reference date
condition 3 = CINstart is the same as CPPend in a group

Then we combine with something like: (condition 1 or condition 2) or condition 3

Reply:
Thanks Will. Have now split up the condition logic to make it more readable. Don't think what I have is incorrect though.